power_grid_model_c/power_grid_model/include/power_grid_model/math_solver/common_solver_functions.hpp

```diff
@@ -59,20 +60,64 @@ template <symmetry_tag sym> inline void copy_y_bus(YBus<sym> const& y_bus, Compl
-        output.source[source].i = dot(y_ref, u_ref - output.u[bus_number]);
-        output.source[source].s = output.u[bus_number] * conj(output.source[source].i);
+                                    PowerFlowInput<sym> const& input, SolverOutput<sym>& output,
+                                    ComplexValue<sym> const& i_load_gen_bus) {
+    ComplexValue<sym> const i_inj_t = conj(output.bus_injection[bus_number] / output.u[bus_number]) - i_load_gen_bus;
+    std::vector<Idx> sources_acc(sources.size());
+    std::ranges::transform(sources, sources_acc.begin(), [&](Idx const source) { return source; });
+    if (sources_acc.empty()) {
+        ; // Do nothing
+    } else if (sources_acc.size() == 1) {
+        output.source[sources_acc[0]].i = i_inj_t;
+        output.source[sources_acc[0]].s = output.u[bus_number] * conj(output.source[sources_acc[0]].i);
+    } else {
+        std::vector<ComplexTensor<sym>> y_ref_acc(sources.size());
+        std::vector<ComplexValue<sym>> i_ref_acc(sources.size());
+        std::ranges::transform(sources, y_ref_acc.begin(), [&](Idx const source) -> ComplexTensor<sym> {
+            return y_bus.math_model_param().source_param[source];
+        });
+        std::ranges::transform(sources, i_ref_acc.begin(), [&](Idx const source) -> ComplexValue<sym> {
+            ComplexValue<sym> const u_ref{input.source[source]};
+            ComplexTensor<sym> const y_ref = y_bus.math_model_param().source_param[source];
+            return dot(y_ref, u_ref);
+        });
+        ComplexTensor<sym> const y_ref_t = std::accumulate(y_ref_acc.begin(), y_ref_acc.end(), ComplexTensor<sym>{});
+        ComplexTensor<sym> z_ref_t{};
+        if constexpr (is_symmetric_v<sym>) {
+            z_ref_t = 1.0 / y_ref_t;
+        } else {
+            DoubleComplex const s = y_ref_t(0); // Diagonal elements
+            DoubleComplex const m = y_ref_t(1); // Off diagonal elements
+            // z_ref_t is obtained from: y_ref_t -> y_012 -> z_012 -> z_ref_t
+            // s_z, m_z are diagonal and off-diagonal elements of z_ref_t
+            DoubleComplex const s_z = (2.0 / (3.0 * (s - m))) + (1.0 / (3.0 * (s + (2.0 * m))));
+            DoubleComplex const m_z = (1.0 / (3.0 * (s + (2.0 * m)))) - (1.0 / (3.0 * (s - m)));
+            z_ref_t = ComplexTensor<sym>{s_z, m_z};
+        }
+        ComplexValue<sym> const i_ref_t = std::accumulate(i_ref_acc.begin(), i_ref_acc.end(), ComplexValue<sym>{});
+        for (size_t i = 0; i < sources.size(); ++i) {
+            Idx const source = sources_acc[i];
+            ComplexValue<sym> const u_ref_t = dot(z_ref_t, i_ref_t);
+            ComplexValue<sym> const u_ref_i{input.source[i]};
+            ComplexValue<sym> delta_u = (u_ref_i - u_ref_t);
+            // Arbitrary precision threshold (20.0) needs further discussion, as well as this whole cutoff.
+            constexpr auto precision = 20.0 * std::numeric_limits<double>::epsilon();
```

Comment:
Make a variable for it:
```suggestion
            // Arbitrary precision threshold (20.0) needs further discussion, as well as this whole cutoff.
            constexpr double theta = 20.0;
```
and then
```suggestion
            constexpr auto precision = theta * std::numeric_limits<double>::epsilon();
```

Reply:
I wanted to leave it as an explicit number such that it would be impossible to miss, at least after the discussion is had. But you are totally right.